thetis/timeintegrator.py

```diff
@@ -85,6 +85,32 @@ def advance_picard(self, t, update_forcings=None, update_lagged=True, update_fie
         """
         raise NotImplementedError(f"Picard iterations are not supported for {self} time integrators.")
 
+    def create_fields_old(self):
+        """
+        Create a copy of fields dictionary to store beginning of timestep values
+        """
+        # create functions to hold the values of previous time step
+        self.fields_old = {}
+        for k in sorted(self.fields):
+            if self.fields[k] is not None:
+                if isinstance(self.fields[k], Function):
+                    self.fields_old[k] = Function(
+                        self.fields[k].function_space())
+                elif isinstance(self.fields[k], Constant):
+                    # Although Constants may be changed by the user, we just
+                    # use the same value here, as assigning to domain-less Constants
+                    # causes issues in the adjoint. Constants with a domain, created by Constaint(.., domain=...),
+                    # are now just Functions on the Real space, so are dealt with under isinstance(..., Function)
+                    self.fields_old[k] = self.fields[k]
+
+    def update_fields_old(self):
+        """
+        Update the values of fields_old with those of fields
+        """
+        for k in sorted(self.fields):
+            if isinstance(self.fields[k], Function):
+                self.fields_old[k].assign(self.fields[k])
```

Comment:
Looks good. It's also a nice refactoring to move this to the base class anyway :)